src/DerivationPath.h

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include <TrustWalletCore/TWCoinType.h>
+#include <TrustWalletCore/TWPurpose.h>
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace TW {
+
+struct DerivationPathIndex {
+    uint32_t value;
+    bool hardened;
+
+    DerivationPathIndex() : value(), hardened(true) {}
+    DerivationPathIndex(uint32_t value, bool hardened = true) : value(value), hardened(hardened) {}
+
+    /// The derivation index.
+    uint32_t derivationIndex() const {
+        if (hardened) {
+            return value | 0x80000000;
+        } else {
+            return value;
+        }
+    }
+
+    std::string string() const {
+        if (hardened) {
+            return std::to_string(value) + "'";
+        } else {
+            return std::to_string(value);
+        }
+    }
+};
+
+/// A BIP32 HD wallet derivation path.
+struct DerivationPath {
+    std::vector<DerivationPathIndex> indices;
+
+    TWPurpose purpose() const {
+        return static_cast<TWPurpose>(indices[0].value);
+    }
+
+    void setPurpose(TWPurpose v) {
+        indices[0] = DerivationPathIndex(v, /* hardened: */true);
+    }
+
+    TWCoinType coin() const {
+        return static_cast<TWCoinType>(indices[1].value);
+    }
+
+    void setCoin(TWCoinType v) {
+        indices[1] = DerivationPathIndex(v, /* hardened: */true);
+    }
+
+    uint32_t account() const {
+        return indices[2].value;
+    }
+
+    void setAccount(uint32_t v) {
+        indices[2] = DerivationPathIndex(v, /* hardened: */true);
+    }
+
+    uint32_t change() const {
+        return indices[3].value;
+    }
+
+    void setChange(uint32_t v) {
+        indices[3] = DerivationPathIndex(v, /* hardened: */false);
+    }
+
+    uint32_t address() const {
+        return indices[4].value;
+    }
+
+    void setAddress(uint32_t v) {
+        indices[4] = DerivationPathIndex(v, /* hardened: */false);
+    }
+
+    DerivationPath() = default;
+    DerivationPath(std::vector<DerivationPathIndex> indices) : indices(indices) {}
+
+    /// Creates a `DerivationPath` by components.
+    DerivationPath(TWPurpose purpose, TWCoinType coin, uint32_t account = 0, uint32_t change = 0, uint32_t address = 0) {
```

Comment:
Can we support different number of levels? e.g. for Stellar (ed25519) path is `m/44'/148'/0'`

Reply:
done